text/0000-reads-clauses-on-methods.md

```diff
@@ -0,0 +1,216 @@
+- Feature Name: reads_clauses_on_methods
+- Start Date: 2021-09-17
+- RFC PR: [dafny-lang/rfcs#0000](https://github.com/dafny-lang/rfcs/pull/0000)
+- Dafny Issue: [dafny-lang/dafny#0000](https://github.com/dafny-lang/dafny/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This proposal simply adds support for optional `reads` clauses on methods, just as they are already supported on functions. While the default for functions is disallowing all reads (i.e. `reads {}`) the default for methods would be allowing all reads (i.e. `reads *`), allowing full backwards-compatibility.
+
+# Motivation
+[motivation]: #motivation
+
+Dafny currently has no direct support for concurrency. The soundness of method verification depends on the assumption that no modification to mutable state occurs other than what is present in the method body. Consider this example of a Dafny class with internal state:
+
+```dafny
+class Counter {
+  var x: int
+  constructor() {
+    x := 0;
+  }
+  function method Get() 
+    reads this
+    ensures Get() == x
+  {
+    x
+  }
+  method Increment() 
+    modifies this
+    ensures x == old(x) + 1
+  {
+    x := x + 1;
+  }
+}
+```
+
+If this class is compiled to C#, for example, and its instances are shared between multiple threads, its verified properties are no longer true. If two threads both invoke `Increment` on a single counter, the effect could easily be to only increment `x` once, since reading and writing `x` does not occur atomically. Even the post-condition on `Get` is no longer valid, because the value of `x` could change in-between reading its value and returning from the compiled body.
+
+This proposal does not aim to add concurrency primitives to Dafny, or even propose any particular choice of paradigm for asynchronous or concurrent execution. Instead, it supports verifying that a particular method is safe to use in a concurrent environment, by verifying that it does not operate on any shared mutable state.
```

Comment:
> by verifying that it does not operate on any shared mutable state.

Can't you also achieve that by controlling which arguments are passed to the method? If the method doesn't have any pointers to traverse from, then it can't read from or write to any mutable state, can it?

Or are we talking about static mutable state?

Reply:
That could work too! But there are a few downsides to that approach:

1. The analysis to traverse all paths from a value to find any potential mutable state doesn't exist as any part of Dafny's implementation already, and would have to deal with some tricky cases where it would have to be overly conservative (e.g. if you have a generic method that takes a `T` value, you'd probably have to be safe and reject it just in case it ends up making mutable state reachable). The framing approach is quite a bit cheaper to implement since `modifies` already exists and `reads` can heavily reuse its verification machinery.
2. That analysis would not be compatible with modular verification, since you'd need to look inside dependent modules in case there is any mutable state lurking there. Whereas reads and modifies frames will be/are part of the specification of module contents.
3. Rejecting any input connected to mutable state is more restrictive than necessary, since there's no danger if you don't actually read or write that state.

Reply:
Ok, thanks !